# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [1]:
import pandas as pd
import sqlite3

import warnings
warnings.simplefilter('ignore')

C:\Users\Will\AppData\Local\Temp\ipykernel_10992\2516648483.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [2]:
gosales_connect = sqlite3.connect("go_sales.sqlite")


Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [3]:
product_df = pd.read_sql_query("SELECT * FROM product", gosales_connect)
product_type_df = pd.read_sql_query("SELECT * FROM product_type", gosales_connect)
product_line_df = pd.read_sql_query("SELECT * FROM product_line", gosales_connect)
sales_staff_df = pd.read_sql_query("SELECT * FROM sales_staff", gosales_connect)
sales_branch_df = pd.read_sql_query("SELECT * FROM sales_branch", gosales_connect)
retailer_site_df = pd.read_sql_query("SELECT * FROM retailer_site", gosales_connect)
country_df = pd.read_sql_query("SELECT * FROM country", gosales_connect)
order_header_df = pd.read_sql_query("SELECT * FROM order_header", gosales_connect)
order_details_df = pd.read_sql_query("SELECT * FROM order_details", gosales_connect)
target_df = pd.read_sql_query("SELECT * FROM SALES_TARGETData", gosales_connect)
returned_item_df = pd.read_sql_query("SELECT * FROM returned_item", gosales_connect)
return_reason_df = pd.read_sql_query("SELECT * FROM return_reason", gosales_connect)

In [4]:
order_details_df

ORDER_DETAIL_CODE ORDER_NUMBER PRODUCT_NUMBER  QUANTITY UNIT_COST  \
0                100000         8462             96        16      16.1   
1                100001         9111             96        20      16.1   
2                100002         8451             96        24      16.1   
3                100003         8453             96        18      16.1   
4                100004         8439             96        20      16.1   
...                 ...          ...            ...       ...       ...   
43058             99995         8459             95       146      2.76   
43059             99996         8457             95       172      2.76   
43060             99997         9267             95       192      2.76   
43061             99998         8441             95       192      2.76   
43062             99999         8440             96        22      16.1   

      UNIT_PRICE UNIT_SALE_PRICE TRIAL879  
0          22.54           22.54        T  
1          22.54           22.54        T  
2          22.54           22.54        T  
3          22.54           22.54        T  
4          22.54           22.54        T  
...          ...             ...      ...  
43058       5.55            5.55        T  
43059       5.55            5.55        T  
43060       5.55            5.55        T  
43061       5.55            5.55        T  
43062      22.54           22.54        T  

[43063 rows x 8 columns]

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [5]:
product_df['PRODUCTION_COST'] = product_df['PRODUCTION_COST'].astype(float)
product_tussen_50_100 = (product_df['PRODUCTION_COST'] > 50) & (product_df['PRODUCTION_COST'] < 100)
product_df_filter = product_df.loc[(product_tussen_50_100), ['PRODUCT_NAME','PRODUCTION_COST']]
product_df_filter

PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [6]:
product_margin_20_60 = (product_df['MARGIN'].astype(float) < .2) | (product_df['MARGIN'].astype(float) > .6)
product_df_margin = product_df.loc[(product_margin_20_60), ['PRODUCT_NAME','MARGIN']]
product_df_margin

PRODUCT_NAME MARGIN
29                 Hibernator Pad    .17
32   Canyon Mule Climber Backpack    .17
100             BugShield Natural    .67
101               BugShield Spray    .67
102         BugShield Lotion Lite     .7
103              BugShield Lotion    .63
105             BugShield Extreme    .63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [7]:
country_franc_as_currency = country_df['CURRENCY_NAME'] == 'francs'
country_franc = country_df.loc[(country_franc_as_currency), ['COUNTRY']].sort_values(by='COUNTRY')
country_franc


COUNTRY
7       Belgium
0        France
19  Switzerland

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [8]:
product_margin_50_introduction = product_df['MARGIN'].astype(float) > .5
product_introduction_date = product_df.loc[(product_margin_50_introduction), ['INTRODUCTION_DATE']].drop_duplicates('INTRODUCTION_DATE')
product_introduction_date

INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
79         26-10-2016
100         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [9]:
retailer_adress2_region_known = retailer_site_df['ADDRESS2'].notna() & retailer_site_df['REGION'].notna()
retailer_adress_city = retailer_site_df.loc[(retailer_adress2_region_known), ['ADDRESS1', 'CITY']]
retailer_adress_city


ADDRESS1           CITY
2            4830 Princeton    Albuquerque
5        1700 George Street  San Francisco
13        1488 Orleans Road     Charleston
28      2103 Ashbury Street    Los Angeles
29           333 Ustaff Way      San Diego
..                      ...            ...
367      10423 Kenwood Road     Cincinnati
370   332 South Main Street        Hanover
371     795 South Jefferson    Springfield
376      702 North Meridian  Oklahoma City
377  1210 Somerville Avenue    Chattanooga

[78 rows x 2 columns]

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [10]:
countries_dollar = (country_df['CURRENCY_NAME'] == 'dollars') | (country_df['CURRENCY_NAME'] == 'new dollar')
dollar_countries = country_df.loc[(countries_dollar), ['COUNTRY']]
dollar_countries

COUNTRY
3          Taiwan
6       Australia
14  United States
15         Canada

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [11]:
retailer_starts_with_d = (retailer_site_df['POSTAL_ZONE'].str[0] == 'D') & (retailer_site_df['ADDRESS2'].notnull())
retailer_d = retailer_site_df.loc[(retailer_starts_with_d, ['ADDRESS1', 'ADDRESS2', 'CITY'])]
retailer_d

ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [12]:
returned_item_df['RETURN_CODE'].count()

706

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [13]:
retailer_site_df['REGION'].count()

251

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [14]:
import numpy as np
max_margin = np.max(product_df['MARGIN'])
min_margin = np.min(product_df['MARGIN'])
avg_margin = round(np.mean(product_df['MARGIN'].astype(float)), 2)
print(max_margin, min_margin, avg_margin)

.7 .17 0.4


Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [15]:
unknown_address2_count = retailer_site_df['ADDRESS2'].isnull().sum()
unknown_address2_count


286

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [16]:
unit_sale = order_details_df[order_details_df['UNIT_SALE_PRICE'] < order_details_df['UNIT_PRICE']]
avg_cost_sale = unit_sale['UNIT_COST'].astype(float).mean()
avg_cost_sale

121.9948118049615

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [17]:
staff_position = sales_staff_df.groupby('POSITION_EN', as_index = False)['SALES_STAFF_CODE'].count()
staff_position

POSITION_EN  SALES_STAFF_CODE
0                Branch Manager                19
1              District Manager                 4
2               General Manager                 1
3  Level 1 Sales Representative                13
4  Level 2 Sales Representative                39
5  Level 3 Sales Representative                24
6              Regional Manager                 2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [18]:
staff_phonenumber = sales_staff_df.groupby('WORK_PHONE', as_index = False)['SALES_STAFF_CODE'].count() 
staff_phonenumber_4 = staff_phonenumber[staff_phonenumber['SALES_STAFF_CODE'] > 4]
staff_phonenumber_4


WORK_PHONE  SALES_STAFF_CODE
4     +(44) 121 3505267                 5
10  +(61) 03 2982  4242                 5
11  +31 (0)20 692 93 94                 5
13    +33 1 68 94 52 20                 5
18     1 (206) 292-0012                 5
19     1 (305) 557-4810                 6
20     1 (310) 281-5722                 5
21     1 (403) 232-5986                 5
22     1 (416) 493-5595                 5
23     1 (617) 268-6754                 6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [19]:
merge_retail = pd.merge(retailer_site_df, country_df, left_on = 'COUNTRY_CODE', right_on = 'COUNTRY_CODE')
dutch_retail = merge_retail[merge_retail['COUNTRY'] == 'Netherlands'][['ADDRESS1', 'CITY']]
dutch_retail

ADDRESS1          CITY
164                   Beets 36         Beets
165               Perestraat 1      Deventer
166  Laan van Meerdervoort 966      Den Haag
167             Kalfjeslaan 47    Amstelveen
168               Startbaan 11    Amstelveen
169      Noorderspoorsingel 28     Groningen
170              Hogehilweg 13     Amsterdam
171             Spoorstraat 43    Varsseveld
173             Westhavenweg 8     Amsterdam
174              Westzeedijk 4        Muiden
175            Gooiergracht 91         Laren
176            Hoofdstraat 150   Pieterburen
177                Olde Diek 2         Laren
178     Amsterdamsestraatweg 8      Maarssen
179        Krayenhoffstraat 11     Amsterdam
180         Overschiestraat 19     Amsterdam
181              Havenbuurt 12        Marken
182               Duivendaal 1    Wageningen
184              Voorstraat 37  Nieuweschans
185         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [20]:
product_merge = pd.merge(product_df, product_type_df, left_on='PRODUCT_TYPE_CODE', how='inner', right_on='PRODUCT_TYPE_CODE')
product_eyewear = product_merge[product_merge['PRODUCT_TYPE_EN'] == 'Eyewear'][['PRODUCT_NAME']]
product_eyewear

PRODUCT_NAME
80      Polar Sun
81      Polar Ice
83   Polar Sports
84     Polar Wave
85  Polar Extreme

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [21]:
branch_manager = sales_staff_df['POSITION_EN'] == 'Branch Manager'
sales_retail_order_merge = pd.merge(retailer_site_df, order_header_df, on='RETAILER_SITE_CODE')
sales_retail_merge = pd.merge(sales_retail_order_merge, sales_staff_df[branch_manager], on='SALES_STAFF_CODE')
branch_manager_address = sales_retail_merge[['ADDRESS1', 'FIRST_NAME', 'LAST_NAME']].drop_duplicates('ADDRESS1')
branch_manager_address


ADDRESS1 FIRST_NAME LAST_NAME
0  Alameda Santos, 9876     Bayard     Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [22]:
staff_order = pd.merge(sales_staff_df, order_header_df, how='left', on='SALES_STAFF_CODE')
position_sell_date = staff_order[staff_order['POSITION_EN'].str.contains('Manager')][['POSITION_EN', 'ORDER_DATE']].drop_duplicates()
position_sell_date

POSITION_EN  ORDER_DATE
421     Branch Manager         NaN
422    General Manager         NaN
1276  Regional Manager         NaN
1643  District Manager         NaN
5325    Branch Manager  2022-09-09
5326    Branch Manager  2022-10-31
5327    Branch Manager  2022-12-13
5328    Branch Manager  2022-03-13
5329    Branch Manager  2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [23]:
product_type_merge = pd.merge(product_df, product_type_df, on='PRODUCT_TYPE_CODE')
product_order_merge = pd.merge(product_type_merge, order_details_df, on='PRODUCT_NUMBER')
quantity_over_750 = product_order_merge[product_order_merge['QUANTITY'] > 750][['PRODUCT_NAME', 'PRODUCT_TYPE_EN']].drop_duplicates('PRODUCT_NAME')
quantity_over_750

PRODUCT_NAME    PRODUCT_TYPE_EN
8616                Star Peg              Tents
15975              Firefly 2           Lanterns
37903        BugShield Spray  Insect Repellents
38313  BugShield Lotion Lite  Insect Repellents
38770       BugShield Lotion  Insect Repellents
39618      BugShield Extreme  Insect Repellents
40754         Sun Shelter 15          Sunscreen
41106         Sun Shelter 30          Sunscreen
41681             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [24]:
product_order = pd.merge(product_df, order_details_df, on='PRODUCT_NUMBER')
product_order_sale = product_order[0.4 < (product_order['UNIT_PRICE'].astype(float) - product_order['UNIT_SALE_PRICE'].astype(float)) / product_order['UNIT_PRICE'].astype(float)][['PRODUCT_NAME']].drop_duplicates('PRODUCT_NAME')
product_order_sale

PRODUCT_NAME
2765   Hailstorm Titanium Woods Set
37446             BugShield Natural
37967               BugShield Spray
38306         BugShield Lotion Lite
38772              BugShield Lotion
39561             BugShield Extreme
40799                Sun Shelter 15
41194                Sun Shelter 30

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [25]:
return_reason_merge = pd.merge(return_reason_df, returned_item_df, on='RETURN_REASON_CODE')
return_details_merge = pd.merge(return_reason_merge, order_details_df, on='ORDER_DETAIL_CODE')
return_90 = return_details_merge[(return_details_merge['RETURN_QUANTITY'].astype(float) / return_details_merge['QUANTITY'].astype(float) > 0.9)][['RETURN_DESCRIPTION_EN']].drop_duplicates()
return_90


RETURN_DESCRIPTION_EN
451   Wrong product ordered
521   Wrong product shipped
583  Unsatisfactory product

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [26]:
product_type_merge = pd.merge(product_type_df, product_df, on='PRODUCT_TYPE_CODE')
product_type_count = product_type_merge.groupby('PRODUCT_TYPE_EN', as_index= False)['PRODUCT_NAME'].count().rename(columns={'PRODUCT_NAME': 'NUMBER'})
product_type_count


PRODUCT_TYPE_EN  NUMBER
0             Binoculars       4
1   Climbing Accessories       7
2           Cooking Gear      10
3                Eyewear       5
4              First Aid       5
5       Golf Accessories       4
6      Insect Repellents       5
7                  Irons       4
8                 Knives       5
9               Lanterns      12
10            Navigation       4
11                 Packs       6
12               Putters       3
13                  Rope       4
14                Safety       4
15         Sleeping Bags       7
16             Sunscreen       5
17                 Tents       6
18                 Tools       6
19               Watches       5
20                 Woods       4

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [27]:
retailer_country_merge = pd.merge(country_df, retailer_site_df, on='COUNTRY_CODE')
retailer_country_count = retailer_country_merge.groupby('COUNTRY', as_index= False)['RETAILER_SITE_CODE'].count().rename(columns={'RETAILER_SITE_CODE': 'NUMBER'})
retailer_country_count

COUNTRY  NUMBER
0        Australia      10
1          Austria      10
2          Belgium      10
3           Brazil       1
4           Canada      31
5            China      10
6          Denmark       6
7          Finland       5
8           France      29
9          Germany      32
10           Italy      12
11           Japan      25
12           Korea      10
13          Mexico      10
14     Netherlands      20
15           Spain       9
16          Sweden      12
17     Switzerland      13
18          Taiwan      10
19  United Kingdom      25
20   United States     101

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [28]:
product_type_merge_order_details = pd.merge(product_type_merge, order_details_df, on='PRODUCT_NUMBER')
cooking_gear_sales = product_type_merge_order_details[product_type_merge_order_details['PRODUCT_TYPE_EN'] == 'Cooking Gear']
cooking_gear_sales.loc[:, ["TOTAAL_VERKOCHTE_HOEVEELHEID"]] = cooking_gear_sales["QUANTITY"].sum()
cooking_gear_sales.loc[:, ["GEMIDDELDE_VERKOOPPRIJS"]] = (cooking_gear_sales["UNIT_PRICE"].astype(float) + cooking_gear_sales["UNIT_SALE_PRICE"].astype(float)).astype(float) / 2
cooking_gear_sales[["PRODUCT_NAME", "GEMIDDELDE_VERKOOPPRIJS", "TOTAAL_VERKOCHTE_HOEVEELHEID"]].drop_duplicates()
cooking_gear_sales.groupby("PRODUCT_NAME")[["TOTAAL_VERKOCHTE_HOEVEELHEID", "GEMIDDELDE_VERKOOPPRIJS"]].first()


TOTAAL_VERKOCHTE_HOEVEELHEID  \
PRODUCT_NAME                                              
TrailChef Canteen                                198676   
TrailChef Cook Set                               198676   
TrailChef Cup                                    198676   
TrailChef Deluxe Cook Set                        198676   
TrailChef Double Flame                           198676   
TrailChef Kettle                                 198676   
TrailChef Kitchen Kit                            198676   
TrailChef Single Flame                           198676   
TrailChef Utensils                               198676   
TrailChef Water Bag                              198676   

                           GEMIDDELDE_VERKOOPPRIJS  
PRODUCT_NAME                                        
TrailChef Canteen                            12.13  
TrailChef Cook Set                           54.09  
TrailChef Cup                                 7.15  
TrailChef Deluxe Cook Set                   123.56  
TrailChef Double Flame                      134.63  
TrailChef Kettle                             12.15  
TrailChef Kitchen Kit                        22.84  
TrailChef Single Flame                       65.23  
TrailChef Utensils                           17.40  
TrailChef Water Bag                           6.03

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

In [29]:
target_sales_merge = pd.merge(target_df, sales_staff_df, on='SALES_STAFF_CODE')
target_sales_branch_merge = pd.merge(target_sales_merge, sales_branch_df, on='SALES_BRANCH_CODE')
target_filter = target_sales_branch_merge["CITY"] == "Amsterdam"
target_sales_branch_merge = target_sales_branch_merge[target_filter]
target_data = target_sales_branch_merge.groupby(["FIRST_NAME", "LAST_NAME"])
target_sales_branch_merge["HOOGSTE_SALES"] = target_data["SALES_TARGET"].transform("max")
target_sales_branch_merge["LAAGSTE_SALES"] = target_data["SALES_TARGET"].transform("min")
target_sales_branch_merge[target_filter][["FIRST_NAME", "LAST_NAME", "LAAGSTE_SALES", "HOOGSTE_SALES"]].drop_duplicates()

FIRST_NAME         LAST_NAME LAAGSTE_SALES HOOGSTE_SALES
5467    Belinda  Jansen-Velasquez          1004           984
5827      Ellen           Shapiro           100           997
5881        Jan         Haverkamp          1001          9987
6184      Maria          Laponder          1000          9993

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [30]:
country_sales = pd.merge(country_df, sales_branch_df, on='COUNTRY_CODE')
country_sales_retailer = pd.merge(country_sales, retailer_site_df, on='COUNTRY_CODE')
country_sales_retailer = country_sales_retailer.rename({"CITY_x" : "klanten", "CITY_y" : "verkoper"}, axis=1)
country_sales_retailer[["COUNTRY", "verkoper", "klanten"]].groupby(["COUNTRY", "verkoper"])[["klanten"]].count()


klanten
COUNTRY       verkoper            
Australia     Brisbane           1
              Cairns             2
              Melbourne          2
              NewCastle          2
              Perth              1
...                            ...
United States Seattle            8
              Springfield        4
              Topeka             4
              Tucson             4
              Washington         4

[229 rows x 1 columns]

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [31]:
output = []

for index, staff_row in sales_staff_df.iterrows():

    sales_staff_code = staff_row["SALES_STAFF_CODE"]
    found = False
    for index, order_row in order_header_df.iterrows():
        if order_row["SALES_STAFF_CODE"] == sales_staff_code:
            found = True
    if not found:
        output.append([staff_row["FIRST_NAME"], staff_row["LAST_NAME"]])


pd.DataFrame(output)

0              1
0     Walter         Taylor
1   Penelope        Tamrine
2      Polly         Linton
3      Gracy        Gellens
4       Else         Mörike
5      Frank          Fuchs
6      Fritz         Hirsch
7       Jörg          Kunze
8      Maria      Iacobucci
9       Kick        Kalkman
10     Karin      Bergström
11     Denis           Pagé
12     Sally          White
13     Frank        Bretton
14   Georges  Saint-Germain
15      Alex      Rodriguez
16  Samantha          Floyd
17     Julie          Olsen
18     Corey         Wright
19   Estelle        Johnson
20      Greg        Belding
21     Julie        Béchard
22     Lucía          Reyna
23    Kazumi        Uragome
24     Alice         Walter

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [32]:
gemiddelde_margin = product_df["MARGIN"].astype(float).mean()
totaal = 0
aantal = 0

for index, product in product_df.iterrows():
    margin = float(product["MARGIN"])
    if(margin < gemiddelde_margin):
        aantal += 1
        totaal += margin

gemiddelde_lage_margin = totaal / aantal
pd.DataFrame([aantal, gemiddelde_lage_margin], ["Aantal", "Gemiddelde"])
            

0
Aantal      59.000000
Gemiddelde   0.292203

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [66]:
product_above_500 = order_details_df["UNIT_PRICE"].astype(float) > 500
filtered_product_above_500 = order_details_df[product_above_500]

merged = pd.merge(filtered_product_above_500, returned_item_df, on=["ORDER_DETAIL_CODE"], how="left")

niet_teruggebracht = []

for _, product in filtered_product_above_500.iterrows():
    if product["PRODUCT_NUMBER"] not in merged["PRODUCT_NUMBER"]:
        niet_teruggebracht.append([product_df[product_df["PRODUCT_NUMBER"] == product["PRODUCT_NUMBER"]][["PRODUCT_NAME"]]])


stuff = pd.DataFrame(niet_teruggebracht, columns=["PRODUCT_NAME"])

stuff.drop_duplicates(subset=["PRODUCT_NAME"], inplace=True)

stuff

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [34]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [35]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 